### ISSUE TITLE
Please provide explanation on learning curve behavior on Training & Testing Data in AL

### ISSUE THREAD

**Issue by DebopriyaGhosh-1990** [webis-de/small-text#76]
I am using small-text for active learning on my kubernetes log dataset. I'm starting with initial 5 samples (picked via "random_initialization_balanced" method ) and and in each run adding another sets of 5 samples with .query() method and updating active-learner. Running the active_learning loop 20 times. This is one set of run. The same sets of operation I am repeating multiple times to test how the learning curve is varying.

**Observation :**
1) In Training data: Each time I'm noticing Training curve dips first then start increasing. In some run Training curve moving up first then dips for sometime and then started increasing. During multiple iteration on the active learning loop one observation is common , i.e. Training curve is going down first and then moving up. 

**Question:**
While fitted on same data, why score for prediction on Training data dips first. During training it sees same data, during prediction on same training set why score is low in initial phase.

2) In Test data: Initial score is low and ranging between 40% to 60% score till 60 samples included via active-learning loop iteration. Then suddenly showing steep increase till 100% score.

**Question:**
Why sudden increment in score in test data.

Is it like Training data first time underfitted and hence training learning curve dips first. During that time learning curve on test data  ranging between 40% to 60% score whithout much increment. Post good amount of informative samples picked by active-learner, the training cure going up and test data curve is seep increased.

**Not sure about this behavior.**

Tried Random forest and ConfidenceEnhancedLinearSVC classifier, behavior is similar.

The below screenshots are the learning curves(accuray/Precision/recall/f1score vs no. of samples) fitted on both training and test data on two separate run. Each run started with initial 5 samples picked by random_initialization_balanced method of small-text. Then in each iteration 5 samples picked with query strategy "PredictionEntropy" and active learning happened up to 20 iteration (105 samples).

**Run 1:** 
![Image](https://github.com/user-attachments/assets/c6f1581b-d5a3-4b09-99b1-fc7d62f5004a)

Run 2:

![Image](https://github.com/user-attachments/assets/42a0f0c3-6cec-4a9b-ab6c-2707291a28bf)

**Please help me understanding this behavior.** I am doing an POC on active-learning , so I need to clarify the concept on the learning behavior.

**Comment by chschroeder** [comment]
Hi @DebopriyaGhosh-1990,

The fluctuations in the beginning don't seem that strange to me. Especially transformer models are known to be unstable when trained with only few training instances. The learning curves you see in papers are usually averaged over multiple iterations. Your curve does not look like you averaged it, so it is not unexpected to see fluctuations. You could try to initialize with more instances to see if this gets better.

Also the `random_initialization_balanced()` gives you a good (class-balanced) starting point, which is likely to get worse in terms of class balance with the first query. 

What I find more concerning is the 100% test accuracy, especially combined with the steep increase. Does this occur with all models? If yes, then it's likely a problem with the data. You should look at your data if there is some unintended overlap between train/dev/test.

**Comment by DebopriyaGhosh-1990** [comment]
Hi @chschroeder ,

Thanks for the quick turn-around.

To answer your queries:
1) (a) I am using sklearn-based model RandomForest & SVM as a classification model. 
    (b) Sentence-transformer model "all-MiniLM-L6-v2" has been used for embedding generation or vectorization after cleaning my 
          log dataset.
=> So, as no transformer models are getting used for Classification, do you think this initial dips is model related.

2) After increasing initial sample size -> some smoothness getting observed.

3) Why does test score stay flat then suddenly jump to 100% ?
=> To answer question related to this, this steeped increase of test accuracy has been observed with both RandomForest & SVM models. To explain about data leakage among train & test set, it could be due to the similar log message pattern exists between train & test set as i am dealing with hdfs and kube log datasets and logs are usually having similar structure or template, hence there is overlap in terms of pattern in logs. may be some instance name, servername, pod name etc. varying by keeping the original static part of the message fixed.

So, according to u this sharp increase of test accuracy to 100% after 60 samples could be due to this data pattern similarity among train & test samples ...correct?

Thank you for your support!

Regards,
Debopriya

**Comment by chschroeder** [comment]
1) Okay, then this is a not a transformer problem at least. Not sure about random forests, but SVMs don't have this weakness and should be fine here.

2) Good. Then I don't see a problem here.

3) Maybe there is a certain pattern that leaks information? I cannot really help here from the distance.

> So, according to u this sharp increase of test accuracy to 100% after 60 samples could be due to this data pattern similarity among train & test samples ...correct?

I would suspect either a problem with the train/test split OR your classification problem is really that easy. I can't say anything here without knowing your problem, but since you mentioned logfiles, this could be a possibility.

Is 100% plausible to you? What is the number of classes for this problem?
